api_fetch.js

```diff
@@ -125,9 +138,104 @@ const modal1 = document.querySelector("#modal__bg");
 const view = new ChallengeListView();
 view.render(challengesDiv);
 
-// Filter Challenges by Type, onsite or online
+//------ Filter Challenges by Type, onsite or online ------
+// Get all the types
+const online = document.querySelector("#online");
+const onsite = document.querySelector("#onsite");
+// event listener for online/onsite button
+online.addEventListener("click", showByType);
+onsite.addEventListener("click", showByType);
+// FUNCTION TO HANDLE FILTERBYTYPE CLICK
+function showByType() {
+  //Get all the cards (challenge elements)
+  const cards = document.querySelectorAll(".challenge");
+  const type = document.querySelectorAll(".cardType");
+  //Get the noMatchError element
+  const noMatchError = document.getElementById("noMatchError");
+  // Initialize a variable to track whether a match is found
+  let foundMatch = false;
+  //works if online is checked and onsite not checked
+  if (online.checked && !onsite.checked) {
+    for (let i = 0; i < cards.length; i++) {
+      let card = type[i].textContent;
+      if (card.toLowerCase().includes("online")) {
+        // Show the current challenge element
+        cards[i].style.display = "";
+
+        // Set foundMatch to true since a match was found
+        foundMatch = true;
+      } else {
+        // Hide the current challenge element if it doesn't match the search box value
+        cards[i].style.display = "none";
+      }
+      // If a match was found
+      if (foundMatch) {
+        // Clear the error message
+        noMatchError.innerHTML = "";
+      } else {
+        // If no matches were found, set the error message
+        noMatchError.innerHTML = "No matching challenges";
+      }
+    }
+    //works if onsite is checked and online not checked
+  } else if (onsite.checked && !online.checked) {
```

Comment:
Notice how each of these blocks look the same internally. The only difference is the if statement inside of the loop. I suspect the code could have been made more concise by trying to reuse the same logic. Either as a function, or by doing something like this:

```js
// No outer if statement, just a single loop
for (let i = 0; i < cards.length; i++) {
  const card = type[i].textContent.toLowerCase();
  let foundMatch = false; // Start false
  if (online.checked && card.includes('online')) {
    // Filtering for online, and card includes online
    foundMatch = true;
  }
  if (onsite.checked && card.includes('onsite')) {
    // Filtering fo onsite, and card includes onsite
    foundMatch = true;
  }

  if (foundMatch) {
    cards[i].style.display = '';
  } else {
    cards[i].style.display = 'block';
  }
}
```